joi/src/types.ts

```diff
@@ -0,0 +1,15 @@
+import {
+  FieldValues,
+  ResolverResult,
+  UnpackNestedValue,
+} from 'react-hook-form';
+import type { AsyncValidationOptions, Schema } from 'joi';
+
+export type Resolver = <T extends Schema>(
```

Comment:
here is the type in the react-hook-form repo: 

I think maybe it's better to leave that at `react-hook-form` and centralize there?

```
export type Resolver<
  TFieldValues extends FieldValues = FieldValues,
  TContext extends object = object
> = (
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>> | ResolverResult<TFieldValues>;
```

Reply:
It's not the same type

From react-hook-form:
```ts
export type Resolver<
  TFieldValues extends FieldValues = FieldValues,
  TContext extends object = object
> = (
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>> | ResolverResult<TFieldValues>;
```

And from resolvers:
```ts
export type Resolver = <T extends Schema>(
  schema: T,
  options?: AsyncValidationOptions,
) => <TFieldValues extends FieldValues, TContext>(
  values: UnpackNestedValue<TFieldValues>,
  context?: TContext,
  validateAllFieldCriteria?: boolean,
) => Promise<ResolverResult<TFieldValues>>;
```

The generic type isn't placed at the same point. This update improve resolver's return type. But I'm open for a better idea 😉

I tried to centralized but it breaks severals things: https://github.com/react-hook-form/react-hook-form/pull/3821

Reply:
all good, I realize that after reviewing the rest of the resolvers.